streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                     }
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```

Comment:
I think there is a small risk that if the lock was kept by someone else for more than rebalanceTimeout but it was released later, then this task will be "lost" as it is not picked by anyone. But I think it will be covered by KIP-129 anyways. cc @mjsax